core/src/core_tests/workflow_tasks.rs

```diff
@@ -2505,43 +2505,121 @@ async fn core_internal_flags() {
+        start_timer_cmd(1, Duration::from_secs(1)),
+    ];
+    let expected_command_types_emitted = Some(vec![
+        CommandType::StartTimer,
+        CommandType::CompleteWorkflowExecution,
+    ]);
+    _do_post_terminal_commands_test(
+        commands_sent_by_lang,
         [ResponseType::ToTaskNum(1), ResponseType::AllHistory],
-        mock_workflow_client(),
-    );
-    mh.completion_mock_fn = Some(Box::new(|c| {
-        // Only the complete execution command should actually be sent
-        assert_eq!(c.commands.len(), 1);
-        Ok(Default::default())
-    }));
+        expected_command_types_emitted,
+        t,
+    )
+    .await
+}
+
+#[tokio::test]
+async fn post_terminal_commands_are_retained_when_replaying_and_flag_set() {
+    // History contains a non-terminal command (N) followed by the terminal
+    // command (T), with the MoveTerminalCommands flag set in the last WFT The
+    // test establishes that, when core replays this history, it is consistent
+    // with lang completing an activation with [T, N].
+    let mut t = TestHistoryBuilder::default();
+    t.add_by_type(EventType::WorkflowExecutionStarted);
+    t.add_full_wf_task();
+    t.add_timer_started("1".to_string());
+    t.add_workflow_execution_completed();
+    t.set_flags_last_wft(&[CoreInternalFlags::MoveTerminalCommands as u32], &[]);
+
+    let commands_sent_by_lang = vec![
+        CompleteWorkflowExecution { result: None }.into(),
+        start_timer_cmd(1, Duration::from_secs(1)),
+    ];
+    let expected_command_types_emitted = None;
+
+    _do_post_terminal_commands_test(
+        commands_sent_by_lang,
+        [ResponseType::AllHistory],
+        expected_command_types_emitted,
+        t,
+    )
+    .await
+}
+
+#[tokio::test]
+async fn post_terminal_commands_are_not_retained_when_replaying_and_flag_not_set() {
+    // History contains the terminal command (T) preceded immediately by
+    // WFTCompleted, i.e. without any intervening non-terminal command (N), and
+    // the MoveTerminalCommands flag is not set. The test establishes that when
+    // core replays this history, it is consistent with lang completing an
+    // activation with [T, N].
+    let mut t = TestHistoryBuilder::default();
+    t.add_by_type(EventType::WorkflowExecutionStarted);
+    t.add_full_wf_task();
+    t.add_workflow_execution_completed();
+
+    let commands_sent_by_lang = vec![
+        CompleteWorkflowExecution { result: None }.into(),
+        start_timer_cmd(1, Duration::from_secs(1)),
+    ];
+    let expected_command_types_emitted = None;
+
+    _do_post_terminal_commands_test(
+        commands_sent_by_lang,
+        [ResponseType::AllHistory],
+        expected_command_types_emitted,
+        t,
+    )
+    .await
+}
+
+async fn _do_post_terminal_commands_test(
+    commands_sent_by_lang: Vec<workflow_command::Variant>,
+    response_types: impl IntoIterator<Item = impl Into<ResponseType>>,
+    expected_command_types: Option<Vec<CommandType>>,
```

Comment:
nit: Doesn't need to be an option, can just be empty vec when `None`

Reply:
I think option is a reasonable choice: `None` means the concept of expected commands is irrelevant, since the worker is purely replaying all history and will never emit a completion; that's as opposed to `Some(empty_vec)` which would mean SDK did a non-replay WFT but replied without any commands (e.g. all coroutines are blocked on a workflow condition; also WFT fail might be represented like that).